fairlearn/reductions/exponentiated_gradient/exponentiated_gradient.py

```diff
@@ -0,0 +1,267 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+This module implements the Lagrangian reduction of fair binary
+classification to standard binary classification.
+
+FUNCTIONS
+exponentiated_gradient -- optimize accuracy subject to fairness constraints
+"""
+
+from __future__ import print_function
+
+import numpy as np
+import pandas as pd
+import functools
+from fairlearn.reductions import moments
+from fairlearn.reductions import ReductionsEstimator
+from ._constants import _ACCURACY_MUL, _REGRET_CHECK_START_T, _REGRET_CHECK_INCREASE_T, \
+    _SHRINK_REGRET, _SHRINK_ETA, _MIN_T, _RUN_LP_STEP, _PRECISION, _INDENTATION
+from ._lagrangian import _Lagrangian
+from collections import namedtuple
+
+print = functools.partial(print, flush=True)
+
+
+def _mean_pred(X, hs, weights):
+    """Return a weighted average of predictions produced by classifiers in hs"""
+    pred = pd.DataFrame()
+    for t in range(len(hs)):
+        pred[t] = hs[t](X)
+    return pred[weights.index].dot(weights)
+
+
+# TODO: decide whether we want a structured object instead of the namedtuple
+column_names = "best_classifier best_gap classifiers weights last_t best_t n_oracle_calls"
+expgrad_result = namedtuple("ExpgradResult", column_names)
+
+
+class ExponentiatedGradientResult:
+    def __init__(self, best_classifier, best_gap, classifiers, weights, last_t, best_t,
+                 n_oracle_calls):
+        self._best_classifier = best_classifier
+        self._best_gap = best_gap
+        self._classifiers = classifiers
+        self._weights = weights
+        self._last_t = last_t
+        self._best_t = best_t
+        self._n_oracle_calls = n_oracle_calls
+
+    def _predict_with_best_classifier(self, X):
+        return self._best_classifier(X)
+
+    def _asdict(self):
+        return {
+            "best_classifier": self._best_classifier,
+            "best_gap": self._best_gap,
+            "classifiers": self._classifiers,
+            "weights": self._weights,
+            "last_t": self._last_t,
+            "best_t": self._best_t,
+            "n_oracle_calls": self._n_oracle_calls
+        }
+
+
+class ExponentiatedGradient(ReductionsEstimator):
+    def __init__(self, estimator, disparity_metric, quality_metric, eps=0.01):
```

Comment:
Why the change from constraints? I suspect that's a much better word, and as I said, I'd like to move GridSearch to using moments rather than metrics.DemographicParity etc. (for starters, DemographicParity isn't really a metric, but a condition).

Reply:
Interesting point. I thought we always provide disparity_metric for every one of our methods (including post processing). Would there otherwise be a guide on how to write constraints?

Reply:
Section 2.1 of the Reductions Approach paper, I believe :-)

The rest of this is very definitely getting into the discussion I was trying to articulate last week. I would say that GridSearch and ExponentiatedGradient are closer (in some sense) than PostProcessing, so getting the first two together is more important than getting all three similar.

Reply:
It's constraints again. Please raise again if it's not as you expect it